### DISCUSSION TITLE
User to pick active roles out of all roles he is assigned?

### DISCUSSION THREAD

**Discussion by aman-godara-ria** [keycloak/keycloak#27287] [Q&A]
I want to allow user to select some roles out of all roles he has. This is so that the tester can test the flow by impersonating as different kinds of user without having to create several different accounts and giving different roles to each account.

Is this possible in keycloak, in other terms what I want is for the access token to be generated with one few roles in it, instead of having all roles of a user in the access token?

**Comment by jbman** [comment]
You can define a client scope which returns only a subset of the user's roles. This is usually used, so that an application gets only those roles in the token, which are relevant for authorization checks. 
Please see: https://www.keycloak.org/docs/23.0.7/server_admin/#_role_scope_mappings

When obtaining the token, the client application could can request on of the optional scopes (via OAuth2 "scope2 parameter in URL), so that the token will contain only the roles configured for this scope. You can test it as described here: https://www.keycloak.org/docs/23.0.7/server_admin/#_client_scopes_evaluate

**Reply by aman-godara-ria** [reply]
The only client scope that allows for role-scope mappings is the dedicate client scope (`Dedicated scope and mappers for this client`). How can I configure role-scope mappings for client scopes that are created by me?

By changing the role-scope mappings for the dedicated client I will end up changing the roles that go into access token for all the users that use that particular client, whereas in my use case I want the tester to dynamically select few roles out of all roles he has (I will give tester all realm roles firstly) so that he can test out different kinds of users (who have only a subset of realm roles).

I hope my understanding after reading these 2 links is right, otherwise feel free to point out my mistakes. @jbman Please let me know if you are not aware of any such solution, so that I can either get the requirements changed or let my team know that this won't be possible. Thanks
